pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ReaderHandler.java

```diff
@@ -0,0 +1,247 @@
+                    topic, e);
+            close(WebSocketError.FailedToSubscribe, e.getMessage());
+        }
+    }
+
+	private void receiveMessage() {
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] [{}] [{}] Receive next message", getSession().getRemoteAddress(), topic, subscription);
+        }
+
+        reader.readNextAsync().thenAccept(msg -> {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] [{}] [{}] Got message {}", getSession().getRemoteAddress(), topic, subscription,
+                        msg.getMessageId());
+            }
+
+            ConsumerMessage dm = new ConsumerMessage();
+            dm.messageId = Base64.getEncoder().encodeToString(msg.getMessageId().toByteArray());
+            dm.payload = Base64.getEncoder().encodeToString(msg.getData());
+            dm.properties = msg.getProperties();
+            dm.publishTime = DATE_FORMAT.format(Instant.ofEpochMilli(msg.getPublishTime()));
+            if (msg.hasKey()) {
+                dm.key = msg.getKey();
+            }
+            final long msgSize = msg.getData().length;
+
+            try {
+                getSession().getRemote()
+                        .sendString(ObjectMapperFactory.getThreadLocal().writeValueAsString(dm), new WriteCallback() {
+                            @Override
+                            public void writeFailed(Throwable th) {
+                                log.warn("[{}/{}] Failed to deliver msg to {} {}", reader.getTopic(), subscription,
+                                        getRemote().getInetSocketAddress().toString(), th.getMessage());
+                                pendingMessages.decrementAndGet();
+                                // schedule receive as one of the delivery failed
+                                service.getExecutor().execute(() -> receiveMessage());
+                            }
+
+                            @Override
+                            public void writeSuccess() {
+                                if (log.isDebugEnabled()) {
+                                    log.debug("[{}/{}] message is delivered successfully to {} ", reader.getTopic(),
+                                            subscription, getRemote().getInetSocketAddress().toString());
+                                }
+                                updateDeliverMsgStat(msgSize);
+                                pendingMessages.getAndDecrement();
+                            }
+                        });
+            } catch (JsonProcessingException e) {
+                close(WebSocketError.FailedToSerializeToJSON);
+            }
+
+            int pending = pendingMessages.incrementAndGet();
+            if (pending < maxPendingMessages) {
+                // Start next read in a separate thread to avoid recursion
+                service.getExecutor().execute(() -> receiveMessage());
+            } else {
+                // Resume delivery
+                receiveMessage();
```

Comment:
@rdhabalia 
Please let me confirm.
I will try to change ``readerHandler`` like the following codes.
Does this match what you mean ?
``` 
                            public void writeSuccess() {
                                 if (log.isDebugEnabled()) {
                                     log.debug("[{}/{}] message is delivered successfully to {} ", reader.getTopic(),
                                             subscription, getRemote().getInetSocketAddress().toString());
                                 }
                                 updateDeliverMsgStat(msgSize);
 -                               pendingMessages.getAndDecrement();
                             }
.
.
.
              int pending = pendingMessages.incrementAndGet();
              if (pending < maxPendingMessages) {
                  // Start next read in a separate thread to avoid recursion
                  service.getExecutor().execute(() -> receiveMessage());
 -            } else {
 -                // Resume delivery
 -                receiveMessage();
              }
.
.
.
   @Override
    public void onWebSocketText(String message) {
        super.onWebSocketText(message);

+        int pending = pendingMessages.getAndDecrement();
+        if (pending >= maxPendingMessages) {
+            // Resume delivery
+            receiveMessage();
+        }
    }
```